File: backend/app/routers/fixed_assets.py
from decimal import Decimal
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.deps import get_current_user
from app.models.auth import User
from app.models.accounting import FixedAsset
from app.models.fixed_asset import DepreciationEntry
from app.schemas.fixed_asset import (
    FixedAssetCreate, FixedAssetUpdate, FixedAssetResponse,
    DepreciationEntryResponse, RunDepreciationRequest, RunDepreciationResponse,
)

router = APIRouter(prefix="/api/fixed-assets", tags=["TSCĐ"])


# ─── Assets ───────────────────────────────────────────────────────────────────

@router.get("", response_model=list[FixedAssetResponse])
def list_assets(
    trang_thai: str | None = Query(None),
    phap_nhan_id: int | None = Query(None),
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    q = db.query(FixedAsset).order_by(FixedAsset.ma_ts)
    if trang_thai:
        q = q.filter(FixedAsset.trang_thai == trang_thai)
    if phap_nhan_id:
        q = q.filter(FixedAsset.phap_nhan_id == phap_nhan_id)
    return q.all()


@router.post("", response_model=FixedAssetResponse, status_code=201)
def create_asset(
    data: FixedAssetCreate,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    asset = FixedAsset(**data.model_dump())
    db.add(asset)
    db.commit()
    db.refresh(asset)
    return asset


@router.get("/{asset_id}", response_model=FixedAssetResponse)
def get_asset(
    asset_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    a = db.get(FixedAsset, asset_id)
    if not a:
        raise HTTPException(404, "Không tìm thấy tài sản")
    return a


@router.patch("/{asset_id}", response_model=FixedAssetResponse)
def update_asset(
    asset_id: int,
    data: FixedAssetUpdate,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    a = db.get(FixedAsset, asset_id)
    if not a:
        raise HTTPException(404, "Không tìm thấy tài sản")
    for k, v in data.model_dump(exclude_none=True).items():
        setattr(a, k, v)
    db.commit()
    db.refresh(a)
    return a


# ─── Depreciation entries ─────────────────────────────────────────────────────

@router.get("/{asset_id}/depreciation", response_model=list[DepreciationEntryResponse])
def list_depreciation(
    asset_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    if not db.get(FixedAsset, asset_id):
        raise HTTPException(404, "Không tìm thấy tài sản")
    return (
        db.query(DepreciationEntry)
        .filter(DepreciationEntry.asset_id == asset_id)
        .order_by(DepreciationEntry.ky)
        .all()
    )


# ─── Run depreciation batch ───────────────────────────────────────────────────

@router.post("/run-depreciation", response_model=RunDepreciationResponse)
def run_depreciation(
    data: RunDepreciationRequest,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    """Chạy khấu hao tháng cho toàn bộ TSCĐ đang dùng, bỏ qua đã KH kỳ này."""
    assets = db.query(FixedAsset).filter(
        FixedAsset.trang_thai == "dang_su_dung",
        FixedAsset.bo_qua_hach_toan == False,
        FixedAsset.da_khau_hao_thang < FixedAsset.so_thang_khau_hao,
    ).all()

    total = Decimal("0")
    count = 0

    for a in assets:
        exists = db.query(DepreciationEntry).filter(
            DepreciationEntry.asset_id == a.id,
            DepreciationEntry.ky == data.ky,
        ).first()
        if exists:
            continue

        kh_thang = (a.nguyen_gia / a.so_thang_khau_hao).quantize(Decimal("1"))
        con_lai = a.nguyen_gia - a.gia_tri_da_khau_hao
        so_tien = min(kh_thang, con_lai)
        if so_tien <= 0:
            continue

        entry = DepreciationEntry(
            asset_id=a.id,
            ky=data.ky,
            so_tien_kh=so_tien,
            gia_tri_da_kh_sau=a.gia_tri_da_khau_hao + so_tien,
        )
        db.add(entry)
        a.da_khau_hao_thang += 1
        a.gia_tri_da_khau_hao += so_tien
        if a.da_khau_hao_thang >= a.so_thang_khau_hao:
            a.trang_thai = "da_kh_het"

        total += so_tien
        count += 1

    db.commit()
    return RunDepreciationResponse(ky=data.ky, so_tscd_da_kh=count, tong_so_tien_kh=total)

File: frontend/src/api/fixedAssets.ts
import client from './client'

export interface FixedAsset {
  id: number
  ma_ts: string
  ten_ts: string
  ngay_mua: string
  nguyen_gia: number
  so_thang_khau_hao: number
  da_khau_hao_thang: number
  gia_tri_da_khau_hao: number
  trang_thai: string
  phan_xuong_id: number | null
  phap_nhan_id: number | null
  tk_nguyen_gia: string
  tk_khau_hao: string
  tk_chi_phi: string
  bo_qua_hach_toan: boolean
  created_at: string
}

export interface DepreciationEntry {
  id: number
  asset_id: number
  ky: string
  so_tien_kh: number
  gia_tri_da_kh_sau: number
  journal_entry_id: number | null
  created_at: string
}

export interface DepreciationReportItem {
  id: number
  ma_ts: string
  ten_ts: string
  ngay_mua: string
  nguyen_gia: number
  gia_tri_da_khau_hao: number
  gia_tri_con_lai: number
  so_thang_khau_hao: number
  da_khau_hao_thang: number
  trang_thai: string
  phan_xuong_id: number | null
  phap_nhan_id: number | null
  tk_khau_hao: string
  tk_chi_phi: string
  so_tien_du_kien: number
  so_tien_da_hach_toan: number
  da_hach_toan: boolean
  journal_entry_id: number | null
}

export interface DepreciationReport {
  ky: string
  tong_du_kien: number
  tong_da_hach_toan: number
  items: DepreciationReportItem[]
}

export const fixedAssetApi = {
  list: (params?: {
    search?: string
    trang_thai?: string
    phap_nhan_id?: number
    phan_xuong_id?: number
  }) => client.get<FixedAsset[]>('/fixed-assets', { params }).then(r => r.data),
  create: (data: Partial<FixedAsset>) => client.post<FixedAsset>('/fixed-assets', data).then(r => r.data),
  update: (id: number, data: Partial<FixedAsset>) => client.patch<FixedAsset>(`/fixed-assets/${id}`, data).then(r => r.data),
  depreciation: (id: number) => client.get<DepreciationEntry[]>(`/fixed-assets/${id}/depreciation`).then(r => r.data),
  runDepreciation: (ky: string) => client.post('/fixed-assets/run-depreciation', { ky }).then(r => r.data),
  runAssetDepreciation: (id: number, ky: string) => client.post(`/fixed-assets/${id}/depreciate`, { ky }).then(r => r.data),
  depreciationReport: (params: {
    ky: string
    trang_thai?: string
    phap_nhan_id?: number
    phan_xuong_id?: number
  }) => client.get<DepreciationReport>('/fixed-assets/depreciation-report', { params }).then(r => r.data),
}
